samza-core/src/main/java/org/apache/samza/storage/SideInputStorageManager.java

```diff
@@ -0,0 +1,372 @@
+   * Gets the last processed offset for the given side input {@link SystemStreamPartition}.
+   *
+   * @param ssp side input system stream partition to get the last processed offset for
+   * @return the last processed offset
+   */
+  public String getLastProcessedOffset(SystemStreamPartition ssp) {
+    return lastProcessedOffsets.get(ssp);
+  }
+
+  /**
+   * Processes the incoming side input message envelope and updates the last processed offset for its SSP.
+   *
+   * @param message incoming message to be processed
+   */
+  public void process(IncomingMessageEnvelope message) {
+    SystemStreamPartition ssp = message.getSystemStreamPartition();
+    String storeName = sspsToStore.get(ssp);
+    SideInputProcessor sideInputProcessor = storeToProcessor.get(storeName);
+
+    KeyValueStore keyValueStore = (KeyValueStore) stores.get(storeName);
+    Collection<Entry<?, ?>> entriesToBeWritten = sideInputProcessor.process(message, keyValueStore);
+    keyValueStore.putAll(ImmutableList.copyOf(entriesToBeWritten));
+
+    // update the last processed offset
+    lastProcessedOffsets.put(ssp, message.getOffset());
+  }
+
+  /**
+   * Writes the offset files for all side input stores one by one. There is one offset file per store.
+   * Its contents are a JSON encoded mapping from each side input SSP to its last processed offset, and a checksum.
+   */
+  private void writeOffsetFiles() {
+    storeToSSps.entrySet().stream()
+        .filter(entry -> isPersistedStore(entry.getKey())) // filter out in-memory side input stores
+        .forEach((entry) -> {
+            String storeName = entry.getKey();
+            Map<SystemStreamPartition, String> offsets = entry.getValue().stream()
+              .filter(lastProcessedOffsets::containsKey)
+              .collect(Collectors.toMap(Function.identity(), lastProcessedOffsets::get));
+
+            try {
+              String checkpoint = OBJECT_MAPPER.writeValueAsString(offsets);
+              File offsetFile = new File(getStoreLocation(storeName), OFFSET_FILE);
+              FileUtil.writeWithChecksum(offsetFile, checkpoint);
+            } catch (Exception e) {
+              LOG.error("Encountered error while checkpointing to the file due to", e);
+              throw new SamzaException("Failed to checkpoint for side input store " + storeName, e);
+            }
+          });
+  }
+
+  /**
+   * Gets the side input SSP offsets for all stores from their local offset files.
+   *
+   * @return a {@link Map} of {@link SystemStreamPartition} to offset in the offset files.
+   */
+  @SuppressWarnings("unchecked")
+  private Map<SystemStreamPartition, String> getFileOffsets() {
+    LOG.info("Loading initial offsets from the file for side input stores.");
+    Map<SystemStreamPartition, String> fileOffsets = new HashMap<>();
+
+    stores.keySet().forEach(storeName -> {
+        LOG.debug("Reading local offsets for store {}", storeName);
+
+        File storeLocation = getStoreLocation(storeName);
```

Comment:
Creating File objects is fairly cheap (it's a regular object). Since it's also an immutable object and may not reflect changes in the underlying file, it's safer to create a new object when needed. Does make sense to share this logic across the two implementations (TaskStorageManager and TaskSideInputStorageManager). Added this to SAMZA-1780.